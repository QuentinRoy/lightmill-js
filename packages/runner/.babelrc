{
  "env": {
    "test": {
      "presets": [
        [
          "env",
          {
            "targets": { "node": "current" }
          }
        ]
      ],
      "plugins": [
        "transform-object-rest-spread",
        "transform-async-generator-functions"
      ]
    },
    "production": {
      "presets": [
        [
          "env",
          {
            "targets": { "browsers": ">1%" },
            "modules": false,
            "loose": true
          }
        ]
      ],
      "plugins": [
        "transform-object-rest-spread",
        [
          "transform-runtime",
          {
            "helpers": false,
            "polyfill": false,
            "regenerator": true
          }
        ],
        "external-helpers"
      ]
    }
  }
}
