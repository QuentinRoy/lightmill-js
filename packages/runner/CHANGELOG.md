# Change Log

## 3.0.0

### Major Changes

- 37f1b03: New Runner API as default export. The old `run` function is still available as a named export.
- a35f3f9: Runner API has changed. Runner is now provided as a single run function, and the store argument has been removed.
- d2f8ed5: Remove Runner as a default export. There is no default export anymore. Runner is provided as a named export. RunnerProps type is now exported.
- aed9788: Update Runner interface.
- 4bbaa8e: Distribute package as ES module only.

### Minor Changes

- 4bbaa8e: Refactor to typescript. The library now provides typescript types.

### Patch Changes

- 2d3d87e: Remove package.json engines directive which fixes a warning when consumer uses a different node version.

## 3.0.0-beta.25

### Patch Changes

- 2d3d87e: Remove package.json engines directive which fixes a warning when consumer uses a different node version.

## 3.0.0-beta.23

### Major Changes

- aed9788: Update Runner interface.

## 3.0.0-alpha.10

### Major Changes

- d2f8ed5: Remove Runner as a default export. There is no default export anymore. Runner is provided as a named export. RunnerProps type is now exported.

## 3.0.0-alpha.9

### Major Changes

- 37f1b03: New Runner API as default export. The old `run` function is still available as a named export.

## 3.0.0-alpha.6

### Major Changes

- a35f3f9: Runner API has changed. Runner is now provided as a single run function, and the store argument has been removed.
- 4bbaa8e: Distribute package as ES module only.

### Minor Changes

- 4bbaa8e: Refactor to typescript. The library now provides typescript types.

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="3.0.0-alpha.4"></a>

# [3.0.0-alpha.4](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/compare/v3.0.0-alpha.3...v3.0.0-alpha.4) (2018-08-20)

### Bug Fixes

- **runner:** fix esm export ([db4595c](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/db4595c))

### Features

- change UMD global exports ([ddbcbd2](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/ddbcbd2))

### BREAKING CHANGES

- Global exports (when the packages are directly install from script tag in the HTML) are now contained in the `lightmill` namespace.

<a name="3.0.0-alpha.3"></a>

# [3.0.0-alpha.3](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/compare/v3.0.0-alpha.2...v3.0.0-alpha.3) (2018-08-20)

### Features

- better esm export support ([d5ecdcb](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/d5ecdcb))
- **runner:** stores' log method now takes the log as first argument ([8c9c518](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/8c9c518))

### BREAKING CHANGES

- **runner:** Store interface has changed

<a name="3.0.0-alpha.2"></a>

# [3.0.0-alpha.2](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/compare/v3.0.0-alpha.1...v3.0.0-alpha.2) (2018-08-17)

### Bug Fixes

- **runner:** fix first task always being skipped ([e5f543c](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/e5f543c))

### Features

- **runner:** new interface for the stores ([b776bbb](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/b776bbb))
- **runner:** refactor runner to remove babel-runtime dependency ([47832c0](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/47832c0))
- **runner:** rename runner.start() to run ([57349c1](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/57349c1))
- **runner:** rename runner's taskManager argument to runTask ([b1fdaf4](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/b1fdaf4))

### BREAKING CHANGES

- **runner:** runner's taskManager argument have been rename to runTask
- **runner:** runner.start has been rename to run.
- **runner:** Stores do not need to implement `log` anymore but should implement `getLogger(logType)` that returns the corresponding logger instead.

<a name="2.0.0-3"></a>

# 2.0.0-3 (2017-08-08)

<a name="2.0.0-2"></a>

# 2.0.0-2 (2017-06-27)

### Bug Fixes

- **lightmill-runner:** `runExperiment` calls `app.crash` then throws if a trial goes wrong. ([7da8203](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/7da8203))
- **run-experiment:** Adapt to new connection interface. ([f4a2297](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/f4a2297))

### Features

- **lightmill-runner:** `runTrials` throws if a post fails. ([cbf61d8](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/cbf61d8))

<a name="2.0.0-1"></a>

# 2.0.0-1 (2017-06-25)

<a name="2.0.0-0"></a>

# 2.0.0-0 (2017-06-25)

<a name="2.0.0-2"></a>

# 2.0.0-2 (2017-06-27)

### Bug Fixes

- **lightmill-runner:** `runExperiment` calls `app.crash` then throws if a trial goes wrong. ([7da8203](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/7da8203))
- **run-experiment:** Adapt to new connection interface. ([f4a2297](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/f4a2297))

### Features

- **lightmill-runner:** `runTrials` throws if a post fails. ([cbf61d8](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection/commit/cbf61d8))

<a name="2.0.0-1"></a>

# 2.0.0-1 (2017-06-25)

<a name="2.0.0-0"></a>

# 2.0.0-0 (2017-06-25)

<a name="2.0.0-1"></a>

# 2.0.0-1 (2017-06-25)

<a name="2.0.0-0"></a>

# 2.0.0-0 (2017-06-25)

<a name="2.0.0-0"></a>

# 2.0.0-0 (2017-06-25)
