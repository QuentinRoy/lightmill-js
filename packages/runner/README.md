# @lightmill/runner

Runs experiments on the web.
