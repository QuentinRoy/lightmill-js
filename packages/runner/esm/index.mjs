export { default } from './runner.mjs';
