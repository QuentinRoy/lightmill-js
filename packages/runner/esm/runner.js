import { reduceAsyncIterator } from './utils';

export default ({ store, taskIterator, taskManager }) => ({
  start: () =>
    reduceAsyncIterator(taskIterator, async (_, task) => {
      // Run the task.
      const measures = await taskManager(task);
      // Log the results.
      await store.log(task.id, { task, measures });
    }).then(async () => {
      // At the very end (when all tasks has been done), we mark the experiment
      // as complete. Do not directly return store.complete because we don't
      // want to return the result of the call.
      await store.complete();
    })
});
