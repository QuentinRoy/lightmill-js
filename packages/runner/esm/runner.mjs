import { asyncForEach } from './utils.mjs';

const Runner = ({ store, taskIterator, runTask }) => ({
  run: () => {
    const taskLogger = store.getLogger('task');
    const runThenLog = task =>
      runTask(task)
        .then(measures => taskLogger.log(task.id, { task, measures }))
        .then(() => {});
    return asyncForEach(taskIterator, runThenLog)
      .then(() => store.complete())
      .then(() => {});
  }
});

export default Runner;
