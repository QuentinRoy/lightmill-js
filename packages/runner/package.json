{
  "name": "@lightmill/runner",
  "version": "3.0.0-alpha.5",
  "description": "JS Runner for LightMill.",
  "main": "dist/lightmill-runner",
  "module": "dist/lightmill-runner.mjs",
  "files": [
    "dist",
    "esm/**/!(*.test|*.spec).?(m)js"
  ],
  "license": "MIT",
  "access": "public",
  "devDependencies": {
    "@babel/core": "7.2.2",
    "@babel/plugin-external-helpers": "7.2.0",
    "@babel/plugin-transform-runtime": "7.2.0",
    "@babel/preset-env": "7.2.3",
    "array-range": "1.0.1",
    "babel-core": "7.0.0-bridge.0",
    "babel-preset-env": "1.7.0",
    "cross-env": "5.2.0",
    "eslint": "5.11.1",
    "eslint-config-airbnb-base": "13.1.0",
    "eslint-plugin-import": "2.14.0",
    "jest": "23.6.0",
    "promise.defer": "1.0.1",
    "rollup": "0.68.2",
    "rollup-plugin-async": "1.2.0",
    "rollup-plugin-babel": "4.1.0",
    "rollup-plugin-commonjs": "9.2.0",
    "rollup-plugin-node-resolve": "4.0.0",
    "wait-then": "0.1.2"
  },
  "scripts": {
    "test": "cross-env NODE_ENV=test jest",
    "coverage": "yarn run test --coverage",
    "build-esm": "node ../../build-tools/esm-reexport.js -s ./esm/index.mjs -o ./dist/lightmill-runner.mjs -nd",
    "build-umd": "cross-env NODE_ENV=production rollup --config rollup.config.js",
    "build": "yarn build-esm; yarn build-umd",
    "prepublish": "npm run build"
  },
  "keywords": [
    "experiment",
    "lightmill"
  ],
  "author": "Quentin Roy <quentin@quentinroy.fr>",
  "repository": "https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-connection",
  "jest": {
    "testMatch": [
      "**/__tests__/**/*.?(m)js?(x)",
      "**/?(*.)(spec|test).?(m)js?(x)"
    ],
    "moduleFileExtensions": [
      "js",
      "json",
      "jsx",
      "node",
      "mjs"
    ],
    "transform": {
      "^.+.m?js$": "babel-jest"
    },
    "collectCoverageFrom": [
      "esm/**/*.?(m)js?(x)"
    ]
  }
}
