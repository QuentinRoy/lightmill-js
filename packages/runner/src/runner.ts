import run, { RunProps, SuperIterator } from './run.js';
export { run, RunProps };

export type BaseTask = { type: string };
export type RunnerProps<Task extends BaseTask> = {
  tasks: {
    [K in Task['type']]: TaskHander<Task & { type: K }>;
  };
  onExperimentStarted?: () => void;
  onLoading?: () => void;
  onExperimentCompleted?: () => void;
};
export type RunnerInstance<Task extends BaseTask> = {
  run: (taskIterator: SuperIterator<Task>) => PromiseLike<void>;
} & RunnerProps<Task>;

export type TaskHander<Task = BaseTask> = (task: Task) => PromiseLike<void>;
export type TaskHandlers<Task extends { type: string }> = {
  [K in Task['type']]: TaskHander<Task & { type: K }>;
};

export class Runner<Task extends BaseTask> {
  #state: 'idle' | 'running' | 'loading' | 'completed' | 'error' = 'idle';
  tasks: TaskHandlers<Task>;
  onExperimentStarted?: () => void;
  onLoading?: () => void;
  onExperimentCompleted?: () => void;

  constructor(props: RunnerProps<Task>) {
    this.tasks = { ...props.tasks };
    this.onExperimentStarted = props.onExperimentStarted;
    this.onLoading = props.onLoading;
    this.onExperimentCompleted = props.onExperimentCompleted;
  }

  get state() {
    return this.#state;
  }

  async #runTask(task: Task) {
    let type: Task['type'] = task.type;
    this.#state = 'running';
    await this.tasks[type](task);
    this.#state = 'loading';
    this.onLoading?.();
  }

  async run(taskIterator: SuperIterator<Task>) {
    try {
      this.#state = 'loading';
      this.onExperimentStarted?.();
      this.onLoading?.();
      await run({ taskIterator, runTask: this.#runTask.bind(this) });
      this.#state = 'completed';
      this.onExperimentCompleted?.();
    } catch (error) {
      this.#state = 'error';
      throw error;
    }
  }
}
