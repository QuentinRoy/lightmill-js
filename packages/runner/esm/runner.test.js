import runner from './runner';

const wait = t =>
  new Promise(resolve => {
    setTimeout(resolve, t);
  });

describe('Runner#start', () => {
  let genTasks;
  let taskManager;
  let logger;

  beforeEach(() => {
    // i is used to make sure all functions are called in the right order,
    // and wait for each others.
    let i = -1;
    const getCallNb = () => {
      i += 1;
      return i;
    };
    genTasks = jest.fn(async function* mockGenTasks() {
      await wait(0);
      yield { id: 'task1', i: getCallNb() };
      await wait(0);
      yield { id: 'task2', i: getCallNb() };
      await wait(0);
      yield { id: 'task3', i: getCallNb() };
      await wait(0);
    });
    taskManager = jest.fn(async task => {
      await wait(0);
      return { result: `result-${task.id}`, i: getCallNb() };
    });
    logger = {
      log: jest.fn(async () => {
        await wait(0);
        return { i: getCallNb() };
      }),
      complete: jest.fn(async () => {
        await wait(0);
        return { i: getCallNb() };
      })
    };
  });

  it('calls the taskManager for each tasks, and log the result', async () => {
    await runner({
      taskIterator: genTasks(),
      store: logger,
      taskManager
    }).start();
    expect(taskManager).toMatchSnapshot();
    expect(logger.log).toMatchSnapshot();
    expect(logger.complete).toMatchSnapshot();
  });
});
