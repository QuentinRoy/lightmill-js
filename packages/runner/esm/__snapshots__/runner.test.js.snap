// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runner#start calls the taskManager for each tasks, and log the result 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "i": 1,
        "id": "task2",
      },
    ],
    Array [
      Object {
        "i": 4,
        "id": "task3",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Runner#start calls the taskManager for each tasks, and log the result 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "task2",
      Object {
        "measures": Object {
          "i": 2,
          "result": "result-task2",
        },
        "task": Object {
          "i": 1,
          "id": "task2",
        },
      },
    ],
    Array [
      "task3",
      Object {
        "measures": Object {
          "i": 5,
          "result": "result-task3",
        },
        "task": Object {
          "i": 4,
          "id": "task3",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`Runner#start calls the taskManager for each tasks, and log the result 3`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
